src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs

```diff
@@ -0,0 +1,441 @@
+            }
+
+            int makeDagTempSlot(TypeWithAnnotations type, BoundDagTemp temp)
+            {
+                object slotKey = (node, temp);
+                return GetOrCreatePlaceholderSlot(slotKey, type);
+            }
+        }
+
+        public override BoundNode VisitSwitchExpression(BoundSwitchExpression node)
+        {
+            // first, learn from any null tests in the patterns
+            int slot = MakeSlot(node.Expression);
+            var expressionState = VisitRvalueWithState(node.Expression);
+            if (slot > 0)
+            {
+                bool foundTopLevelNullCheck = false;
+                var originalInputType = node.Expression.Type;
+                foreach (var arm in node.SwitchArms)
+                {
+                    foundTopLevelNullCheck |= LearnFromAnyNullPatterns(slot, originalInputType, arm.Pattern);
+                }
+
+                if (foundTopLevelNullCheck)
+                    expressionState = new TypeWithState(expressionState.Type, NullableFlowState.MaybeNull);
+            }
+
+            var labelStateMap = LearnFromDecisionDag(node.Syntax, node.DecisionDag, node.Expression, expressionState, ref this.State);
+            var endState = UnreachableState();
+
+            if (!node.ReportedNotExhaustive && node.DefaultLabel != null &&
+                labelStateMap.TryGetValue(node.DefaultLabel, out var defaultLabelState) && defaultLabelState.believedReachable)
+            {
+                SetState(defaultLabelState.state);
+                ReportSafetyDiagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustiveForNull, ((SwitchExpressionSyntax)node.Syntax).SwitchKeyword.GetLocation());
+            }
+
+            foreach (var arm in node.SwitchArms)
+            {
+                SetState(!arm.Pattern.HasErrors && labelStateMap.TryGetValue(arm.Label, out var labelState) ? labelState.state : UnreachableState());
+                VisitRvalue(arm.Value);
+                Join(ref endState, ref this.State);
+            }
+
+            SetState(endState);
+
+            // https://github.com/dotnet/roslyn/issues/34233
+            // We need to recompute the result type and state of the switch expression based on
+            // the result type and state of all of the arms. This can be done in a way similar
+            // to how it is done for an implicit array creation expression.
+            this.ResultType = TypeWithAnnotations.Create(node.Type).ToTypeWithState();
+            return null;
+        }
+
+        public override BoundNode VisitIsPatternExpression(BoundIsPatternExpression node)
+        {
+            Debug.Assert(!IsConditionalState);
+            LearnFromAnyNullPatterns(node.Expression, node.Pattern);
+            var expressionState = VisitRvalueWithState(node.Expression);
+            var labelStateMap = LearnFromDecisionDag(node.Syntax, node.DecisionDag, node.Expression, expressionState, ref this.State);
```

Comment:
Not sure if there is an ordering issue with these two statements.  If we learn first, we might be using the wrong field symbols.  If we visit first, we might have the wrong null states.

Reply:
This ordering is correct: slots are not affected by the recomputed result of visiting.

---
In reply to: [266731119](https://github.com/dotnet/roslyn/pull/34234#discussion_r266731119) [](ancestors = 266731119)